File: combat/src/classicthunder/combat/pile/DrawPile.java
package classicthunder.combat.pile;

import classicthunder.combat.card.CardActor;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class DrawPile {

    private final List<CardActor> cards = new ArrayList<>();
    private final PileFunction pileFunction;

    public DrawPile(PileFunction onUpdate) {

        this.pileFunction = onUpdate;
    }

    public void ClearCards() {

        this.cards.clear();

        pileFunction.onCardCountChanged(0);
    }

    public int GetCardCount() {

        return this.cards.size();
    }

    public void SetPile(List<CardActor> cards) {

        this.cards.clear();

        List<CardActor> tempCards = new ArrayList<>(cards);
        Collections.shuffle(tempCards);
        this.cards.addAll(tempCards);

        pileFunction.onCardCountChanged(this.cards.size());
    }

    public CardActor DrawTopCard() {

        CardActor topCard = this.cards.remove(0);

        pileFunction.onCardCountChanged(this.cards.size());

        return topCard;
    }
}

File: combat/src/classicthunder/combat/pile/DrawPile.kt
package classicthunder.combat.pile

import classicthunder.combat.card.CardActor

internal class DrawPile(private val pileFunction: PileFunction) {
    private var cards: MutableList<CardActor> = ArrayList()

    fun clearCards() {
        cards.clear()
        pileFunction.onCardCountChanged(0)
    }

    fun getCardCount(): Int {
        return cards.size
    }

    fun setPile(cards: List<CardActor>) {
        this.cards = cards.shuffled().toMutableList()
        pileFunction.onCardCountChanged(this.cards.size)
    }

    fun drawTopCard(): CardActor {
        val topCard = cards.removeFirst()
        pileFunction.onCardCountChanged(cards.size)
        return topCard
    }
}